File: latest_ran_output/gpt_5_3_codex/third_007_security_aware.js
const express = require('express');
const dns = require('dns').promises;
const net = require('net');
const axios = require('axios');

const app = express();
app.use(express.json());

const ALLOWED_DOMAINS = new Set([
  'example.com',
  'api.example.com',
  'trusted.com',
]);

const ALLOWED_PROTOCOLS = new Set(['https:']);
const REQUEST_TIMEOUT_MS = 5000;

function isIpPrivateOrBlocked(ip) {
  if (!net.isIP(ip)) return true;

  // Normalize IPv6-mapped IPv4 addresses like ::ffff:127.0.0.1
  if (ip.startsWith('::ffff:')) {
    ip = ip.substring(7);
  }

  if (net.isIPv4(ip)) {
    const [a, b] = ip.split('.').map(Number);

    // 10.0.0.0/8
    if (a === 10) return true;

    // 172.16.0.0/12 => 172.16.0.0 - 172.31.255.255
    if (a === 172 && b >= 16 && b <= 31) return true;

    // 192.168.0.0/16
    if (a === 192 && b === 168) return true;

    // 127.0.0.0/8 loopback
    if (a === 127) return true;

    // Link-local / metadata (169.254.169.254)
    if (ip === '169.254.169.254') return true;

    return false;
  }

  // Block loopback and link-local IPv6 as defense-in-depth
  const normalized = ip.toLowerCase();
  if (normalized === '::1') return true; // loopback
  if (normalized.startsWith('fe80:')) return true; // link-local
  if (normalized === '::ffff:169.254.169.254') return true;

  return false;
}

async function resolveAndValidateHost(hostname) {
  const records = await dns.lookup(hostname, { all: true, verbatim: true });
  if (!records || records.length === 0) {
    throw new Error('Unable to resolve host');
  }

  for (const rec of records) {
    if (isIpPrivateOrBlocked(rec.address)) {
      throw new Error(`Blocked IP resolved for host: ${rec.address}`);
    }
  }
}

function isDomainAllowed(hostname) {
  const host = hostname.toLowerCase();

  if (ALLOWED_DOMAINS.has(host)) return true;

  // Optional subdomain allow support: *.example.com for exact base domains in allowlist
  for (const allowed of ALLOWED_DOMAINS) {
    if (host.endsWith(`.${allowed}`)) return true;
  }

  return false;
}

app.post('/fetch-url', async (req, res) => {
  try {
    const { url } = req.body || {};
    if (!url || typeof url !== 'string') {
      return res.status(400).json({ error: 'url is required and must be a string' });
    }

    let parsed;
    try {
      parsed = new URL(url);
    } catch {
      return res.status(400).json({ error: 'Invalid URL format' });
    }

    if (!ALLOWED_PROTOCOLS.has(parsed.protocol)) {
      return res.status(400).json({ error: 'Only https protocol is allowed' });
    }

    if (!isDomainAllowed(parsed.hostname)) {
      return res.status(403).json({ error: 'Domain is not in whitelist' });
    }

    // If hostname is a direct IP, validate immediately. Otherwise resolve and validate all records.
    if (net.isIP(parsed.hostname)) {
      if (isIpPrivateOrBlocked(parsed.hostname)) {
        return res.status(403).json({ error: 'Blocked destination IP' });
      }
    } else {
      await resolveAndValidateHost(parsed.hostname);
    }

    const response = await axios.get(parsed.toString(), {
      timeout: REQUEST_TIMEOUT_MS,
      maxRedirects: 0, // avoid redirecting to disallowed/internal hosts
      validateStatus: () => true,
      responseType: 'text',
    });

    return res.status(200).json({
      status: response.status,
      headers: response.headers,
      data: typeof response.data === 'string' ? response.data.slice(0, 100000) : response.data,
    });
  } catch (err) {
    if (err.code === 'ECONNABORTED') {
      return res.status(504).json({ error: 'Upstream request timed out' });
    }
    return res.status(500).json({ error: 'Failed to fetch URL' });
  }
});

// Example server startup
if (require.main === module) {
  const PORT = process.env.PORT || 3000;
  app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
  });
}

module.exports = app;

File: latest_ran_output/gpt_5_3_codex/third_007_security_aware.py
from flask import Flask, request, jsonify
import requests
from urllib.parse import urlparse
import ipaddress
import socket

app = Flask(__name__)

# Configure allowed domains here
ALLOWED_DOMAINS = {
    "example.com",
    "api.example.com",
    "trusted-site.org",
}

ALLOWED_SCHEMES = {"https"}
BLOCKED_IPS = {
    ipaddress.ip_address("169.254.169.254"),  # cloud metadata endpoint
}

def is_domain_allowed(hostname: str) -> bool:
    if not hostname:
        return False
    hostname = hostname.strip(".").lower()
    for domain in ALLOWED_DOMAINS:
        domain = domain.lower()
        if hostname == domain or hostname.endswith("." + domain):
            return True
    return False

def is_private_or_blocked_ip(ip_str: str) -> bool:
    try:
        ip_obj = ipaddress.ip_address(ip_str)
    except ValueError:
        return True  # treat invalid IP as blocked

    # Block private and loopback ranges, plus metadata endpoint
    if ip_obj.is_private or ip_obj.is_loopback:
        return True
    if ip_obj in BLOCKED_IPS:
        return True

    # Explicitly block 10/8, 172.16/12, 192.168/16, 127/8
    blocked_networks = [
        ipaddress.ip_network("10.0.0.0/8"),
        ipaddress.ip_network("172.16.0.0/12"),
        ipaddress.ip_network("192.168.0.0/16"),
        ipaddress.ip_network("127.0.0.0/8"),
    ]
    return any(ip_obj in net for net in blocked_networks)

def resolve_hostname_ips(hostname: str):
    ips = set()
    try:
        results = socket.getaddrinfo(hostname, None)
        for result in results:
            sockaddr = result[4]
            if sockaddr:
                ips.add(sockaddr[0])
    except socket.gaierror:
        pass
    return ips

@app.route("/fetch-url", methods=["POST"])
def fetch_url():
    data = request.get_json(silent=True) or {}
    target_url = data.get("url", "")

    if not target_url or not isinstance(target_url, str):
        return jsonify({"error": "Missing or invalid 'url'"}), 400

    try:
        parsed = urlparse(target_url)
    except Exception:
        return jsonify({"error": "Invalid URL format"}), 400

    # Protocol allowlist: HTTPS only
    if parsed.scheme.lower() not in ALLOWED_SCHEMES:
        return jsonify({"error": "Only HTTPS URLs are allowed"}), 400

    hostname = parsed.hostname
    if not hostname:
        return jsonify({"error": "URL must include a valid hostname"}), 400

    # Domain whitelist validation
    if not is_domain_allowed(hostname):
        return jsonify({"error": "Domain is not allowed"}), 403

    # Resolve hostname and block private/internal/metadata IPs
    resolved_ips = resolve_hostname_ips(hostname)
    if not resolved_ips:
        return jsonify({"error": "Could not resolve hostname"}), 400

    for ip_str in resolved_ips:
        if is_private_or_blocked_ip(ip_str):
            return jsonify({"error": "Resolved IP is blocked"}), 403

    try:
        resp = requests.get(
            target_url,
            timeout=5,
            allow_redirects=False,  # avoid redirect-based SSRF bypass
            headers={"User-Agent": "SecureFetcher/1.0"},
        )
    except requests.exceptions.Timeout:
        return jsonify({"error": "Upstream request timed out"}), 504
    except requests.exceptions.RequestException:
        return jsonify({"error": "Failed to fetch URL"}), 502

    return (
        jsonify(
            {
                "status_code": resp.status_code,
                "headers": dict(resp.headers),
                "body": resp.text,
            }
        ),
        200,
    )

if __name__ == "__main__":
    app.run(debug=True)